drv/ignition-api/src/lib.rs

```diff
@@ -33,105 +36,395 @@ pub const PORT_MAX: usize = 40;
+        self.controller.presence_summary()
+    }
+
+    /// Return the raw state for the given port. The internals of `PortState`
+    /// are purposefully opague since they are closely tied to the RTL
+    /// implementation and as such are not intended for direct consumption. The
+    /// `port(..)` and `target(..)` functions below are probably what you want
+    /// to use.
+    #[inline]
+    pub fn raw_port_state(&self, port: u8) -> Result<PortState, IgnitionError> {
+        self.controller.port_state(port)
+    }
+
+    /// Return the state for the given port.
+    pub fn port(&self, port: u8) -> Result<Port, IgnitionError> {
+        self.raw_port_state(port).map(Port::from)
+    }
+
+    /// Return the `Target` for a given port if present.
+    pub fn target(&self, port: u8) -> Result<Option<Target>, IgnitionError> {
+        self.raw_port_state(port)
+            .map(|state| Port::from(state).target)
+    }
+
+    /// Send the given system power `Request` to the given port. Once a request
+    /// is sent and accepted by the Target, it enforces a cooldown before
+    /// subsequent requests are accepted and processed. `SystemPowerOff` and
+    /// `SystemPowerOn` requests have a three second cooldown while a
+    /// `SystemPowerReset` has a six seconds cooldown.
+    #[inline]
+    pub fn send_request(
+        &self,
+        port: u8,
+        request: Request,
+    ) -> Result<(), IgnitionError> {
+        self.controller.send_request(port, request)
+    }
+
+    /// Return the `Counters` for the given port. This function has the
+    /// side-effect of clearing the counters.
+    #[inline]
+    pub fn counters(&self, port: u8) -> Result<Counters, IgnitionError> {
+        self.controller.counters(port)
+    }
+
+    /// Return the `LinkEvents` for the given transceiver and port. See
+    /// `TransceiverSelect` for more details.
+    pub fn link_events(
+        &self,
+        port: u8,
+        txr: TransceiverSelect,
+    ) -> Result<LinkEvents, IgnitionError> {
+        self.controller.link_events(port, txr).map(LinkEvents::from)
+    }
+
+    /// Clear the link events for the given transceiver and port. See
+    /// `TransceiverSelect` for more details.
+    pub fn clear_link_events(
+        &self,
+        port: u8,
+        txr: TransceiverSelect,
+    ) -> Result<(), IgnitionError> {
+        self.controller.clear_link_events(port, txr)
+    }
+
+    /// Return the state of all ports in a single operation. Be aware that this
+    /// reply is fairly large and may require enlarging the stack of the caller.
+    pub fn all_ports(
+        &self,
+    ) -> Result<[Port; PORT_MAX as usize], IgnitionError> {
+        Ok(self.controller.all_port_state()?.map(Port::from))
+    }
+
+    /// Return the `LinkEvents` for all ports and transceivers in a single
+    /// operation.
+    pub fn all_link_events(
+        &self,
+    ) -> Result<[[LinkEvents; 3]; PORT_MAX as usize], IgnitionError> {
```

Comment:
`LinkEvents` doesn't have a field on it that indicates which transceiver it was for (and I think that's correct!); does that mean the caller has to infer the order of the returned `[LinkEvents; 3]` to tell which goes with which receiver? Would it make sense to switch this to a 3-field struct instead, something like

```rust
pub struct AllLinkEvents {
    pub controller: LinkEvents,
    pub target_link0: LinkEvents,
    pub target_link1: LinkEvents,
}
```

It's the same size/layout as `[LinkEvents; 3]` but maybe easier to understand.

Reply:
I came to the same conclusion and reworked this a bit:

`LinkEvents` are now `TransceiverEvents`, which is much closer to what they are (since these are observed by a transceiver). And from there the set of transceiver events observed by a port are now `LinkEvents`, with a name for each event set.

Apologies as I struggle through a consistent naming convention for the thing I built :)